File: Assets/Samples/TSBehaviour/Script/Example4.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Threading.Tasks;
using UnityEngine;
using Puerts;
using Puerts.TSLoader;
using UnityEngine.Serialization;
using UnityEngine.UI;
using Random = System.Random;

namespace TsSamples
{
    public class Example4 : MonoBehaviour
    {
        private JsEnv _mJsEnv;
        [FormerlySerializedAs("ModuleName")] public string moduleName; //可配置加载的js模块 
        
        public Button _btn;
        private void OnEnable()
        {
            if (_mJsEnv == null)
            {
                string path = "D:/PuerTs/TestPuerTs/TsProject/output/";
                _mJsEnv = new JsEnv(new DefaultLoader(path), 8080);
            }
            
            _mJsEnv.ExecuteModule(moduleName + ".mjs");
            Application.runInBackground = true;
        }

        public static async Task RotateForSeconds(GameObject go, float duration)
        {
            var endTime = Time.time + duration;
            while (Time.time < endTime)
            {
                go.transform.Rotate(Vector3.one * Time.deltaTime * 150);
                await Task.Yield();
            }
        }
        
        void Update()
        {
            if (_mJsEnv != null)
                _mJsEnv.Tick();
        }
        
    }
}

File: TsProject/src/Example4.ts
console.log("Example4");

let cube1 = CS.UnityEngine.GameObject.Find("Cube1");
let cube2 = CS.UnityEngine.GameObject.Find("Cube2");
let example4Mono = CS.UnityEngine.GameObject.FindObjectOfType(puerts.$typeof(CS.TsSamples.Example4)) as CS.TsSamples.Example4;

let clickCall = async function()
{
    await puerts.$promise(CS.TsSamples.Example4.RotateForSeconds(cube1,1));
    await puerts.$promise(CS.TsSamples.Example4.RotateForSeconds(cube2,2));
} 

//example4Mono._btn.onClick.AddListener(clickCall);
example4Mono._btn.onClick.AddListener(async ()=>
{
    //await puerts.$promise(CS.TsSamples.Example4.RotateForSeconds(cube1,1));
    //await WaitTimeForSeconds(1);
    //await puerts.$promise(CS.TsSamples.Example4.RotateForSeconds(cube2,2));
    await RotateForSeconds(cube1,1);
    await WaitTimeForSeconds(1);
    await RotateForSeconds(cube2,2);
});

async function WaitTimeForSeconds(sec:number)
{
    console.log("Ts WaitTimeForSeconds:"+sec);
    let promise = new Promise<void>((resolve,reject)=>
    {
        setTimeout(()=>
        {
            resolve();
        },sec * 1000);
    });
    return promise;
}

async function RotateForSeconds(go:CS.UnityEngine.GameObject,d:number):Promise<void>
{
    let endTime = d + CS.UnityEngine.Time.time;
    while(CS.UnityEngine.Time.time < endTime)
    {
        go.transform.Rotate(CS.UnityEngine.Vector3.op_Multiply(CS.UnityEngine.Vector3.one,CS.UnityEngine.Time.deltaTime * 100));
        await puerts.$promise(<any>CS.System.Threading.Tasks.Task.Yield());
    }
}